src/api/operations.ts

```diff
@@ -236,13 +238,13 @@ export const drawEllipsePath = (config: {
   const ym = y + yScale;
 
   return [
-    pushGraphicsState(),
+    translate(centerX, centerY),
```

Comment:
Can we leave the `pushGraphicsState()` and `popGraphicsState()` operators in place? If not the graphics state can be unintentionally altered even after this function returns.

Reply:
I thought the `drawEllipsePath` should be called internally only. If a "user" wants to draw an ellipse, they'll use `drawEllipse` anyway, which already has push, and pop for graphicsstate. I've removed the "export" from `drawEllipsePath` to make sure it is only called within the `operations.ts`. If I re-add the push and pop, it breaks functionality: ellipses are drawn to a different location than you'd expect.

Although I'm not 100% sure of the functionalities / effects of graphicsstates, I think it wouldn't make sense to push 2 graphicsstates for drawing one ellipse. We use only one push, and one pop for all the rest of the draw functions (svgPath, rectangle, image, etc). I'm starting to think that maybe this double-push-pop of graphic states was the reason why rotation didn't work for the first time.

Reply:
I'm afraid we can't remove the `export` for this (or any) function since releases have already been cut that include the `export`. Removing the `export` would be a breaking change, and only large refactorings or big feature additions warrant something like that (although little changes can be included in those releases), because they force users to refactor their code.

That's actually another reason why we have to leave the push and popping operators in. Removing them would also constitute a breaking change that could mess up the code for users that might be calling `drawEllipsePath`. Though, aside from that, there's nothing wrong with having multiple pushs/pops in a given graphics operation. It's actually a fairly common thing with the kind of state-based graphics model that PDF defines.

I tried adding them back to see what would happen and did not notice a difference in the output PDF. Are you sure that adding them back messes up the rotation? Perhaps that was an issue in a previous version of the code you are thinking of?

Reply:
I've just created 2 screenshots. The left one is without the additional "push-pop", the right one is with the additional "push-pop" of graphicsstates inside "drawEllipsePath". That's the only difference in the code. So, as you can see, it basically breaks the functionality of drawing ellipses as we know it.

![test1](https://user-images.githubusercontent.com/10392261/87218235-2a706e00-c351-11ea-9319-6339b700c030.PNG)


Additionally: I don't know why anyone would use "drawEllipsePath" by itself, as it's clearly just part of the job when you want to draw an ellipse. I'd bet that all of the users who draw ellipses on PDFs with this lib use "drawEllipse", and not "drawEllipsePath". If (for any reason) someone really uses "drawEllipsePath" at the moment, they could just refactor that to "drawEllipse", and it would do the same thing. Except they'd have the opportunity to add additional (but optional) parameters (like colors, opacity, dashed borders, etc). So, I get that it's a breaking change, but it'd be super easy to solve the issues (even easier if the user reads the changelog, and the "migration guide" - if we're about to have one).

I probably can't see the bigger picture, but I just can't see why this is a problem.

Also, if we take a look at "svgPathToOperators", it doesn't "push-pop" graphicsstates either, but it is an exported function. that calls graphics-related operators (like `appendBezierCurve`, `moveTo`, `lineTo`, etc). Why does "drawEllipsePath" need to have "push-pop", if "svgPathToOperators" doesn't need them? To me, it doesn't really make any sense, but again: I probably can't see the bigger picture.

Reply:
Yep, that certainly looks broken to me! I'm still not able to reproduce what you've captured above though. I made the following changes:
```diff
diff --git a/src/api/operations.ts b/src/api/operations.ts
index 177ef22..d24f55d 100644
--- a/src/api/operations.ts
+++ b/src/api/operations.ts
@@ -220,8 +220,7 @@ const drawEllipsePath = (config: {
   xScale: number | PDFNumber;
   yScale: number | PDFNumber;
   rotate?: Rotation;
-}): PDFOperator[] =>
-{
+}): PDFOperator[] => {
   const centerX = asNumber(config.x);
   const centerY = asNumber(config.y);
   const xScale = asNumber(config.xScale);
@@ -238,6 +237,7 @@ const drawEllipsePath = (config: {
   const ym = y + yScale;
 
   return [
+    pushGraphicsState(),
     translate(centerX, centerY),
     rotateRadians(toRadians(config.rotate ?? degrees(0))),
     moveTo(x, ym),
@@ -245,6 +245,7 @@ const drawEllipsePath = (config: {
     appendBezierCurve(xm + ox, y, xe, ym - oy, xe, ym),
     appendBezierCurve(xe, ym + oy, xm + ox, ye, xm, ye),
     appendBezierCurve(xm - ox, ye, x, ym + oy, x, ym),
+    popGraphicsState(),
   ];
 };
 
@@ -287,7 +288,6 @@ export const drawEllipse = (options: {
     popGraphicsState(),
   ].filter(Boolean) as PDFOperator[];
 
-
 export const drawSvgPath = (
   path: string,
   options: {
```
And when I draw an ellipse with the following:
```js
  const pdfDoc = await PDFDocument.create();

  const page = pdfDoc.addPage([300, 300]);
  page.drawEllipse({
    x: 100,
    y: 100,
    yScale: 75,
    xScale: 50,
    color: rgb(1, 0, 0),
    rotate: degrees(45),
  });

  const pdfBytes = await pdfDoc.save();
```
the document looks the same whether I add the push/pop operators or not.

But I don't mean to be a stickler and I could certainly just be doing something wrong on my end. What we can do is just create a new function that is not exported and deprecate `drawEllipsePath` (with an `/** @deprecated */` comment). And the new function would just be the version of `drawEllipsePath` you've implemented in this PR.

Reply:
@Hopding Have you tried test1 (from the integration tests)? That's the one I'm testing with. When I make the same changes you showed me in your last comment (pushGraphicsState, popGraphicsState), it has the effect of what you can see on my screenshot above. Anyway, you're suggestion works for me as well. I've just pushed a PR with it. Thank you for your flexibility!